--- a/llvm/lib/Support/Path.cpp
+++ b/llvm/lib/Support/Path.cpp
@@ -884,9 +884,11 @@
   auto EC = createUniqueFile(Model, FD, ResultPath, OF_None, Mode);
   if (EC)
     return EC;
+#if !defined(__wasm__)
   // FD is only needed to avoid race conditions. Close it right away.
   close(FD);
+#endif
   return EC;
 }
 
@@ -930,9 +932,11 @@
   auto EC = createTemporaryFile(Prefix, Suffix, FD, ResultPath, Flags);
   if (EC)
     return EC;
+#if !defined(__wasm__)
   // FD is only needed to avoid race conditions. Close it right away.
   close(FD);
+#endif
   return EC;
 }
 
@@ -1028,6 +1032,11 @@
 
 #ifndef __APPLE__
 std::error_code copy_file(const Twine &From, const Twine &To) {
+#if defined(__wasm__)
+  (void)From;
+  (void)To;
+  return std::make_error_code(std::errc::operation_not_supported);
+#else
   int ReadFD, WriteFD;
   if (std::error_code EC = openFileForRead(From, ReadFD, OF_None))
     return EC;
@@ -1043,10 +1052,16 @@
   close(WriteFD);
 
   return EC;
+#endif
 }
 #endif
 
 std::error_code copy_file(const Twine &From, int ToFD) {
+#if defined(__wasm__)
+  (void)From;
+  (void)ToFD;
+  return std::make_error_code(std::errc::operation_not_supported);
+#else
   int ReadFD;
   if (std::error_code EC = openFileForRead(From, ReadFD, OF_None))
     return EC;
@@ -1056,11 +1071,17 @@
   close(ReadFD);
 
   return EC;
+#endif
 }
 
 ErrorOr<MD5::MD5Result> md5_contents(int FD) {
   sandbox::violationIfEnabled();
 
+#if defined(__wasm__)
+  (void)FD;
+  return std::make_error_code(std::errc::operation_not_supported);
+#else
   MD5 Hash;
 
   constexpr size_t BufSize = 4096;
@@ -1078,11 +1099,17 @@
   MD5::MD5Result Result;
   Hash.final(Result);
   return Result;
+#endif
 }
 
 ErrorOr<MD5::MD5Result> md5_contents(const Twine &Path) {
   sandbox::violationIfEnabled();
 
+#if defined(__wasm__)
+  (void)Path;
+  return std::make_error_code(std::errc::operation_not_supported);
+#else
   int FD;
   if (auto EC = openFileForRead(Path, FD, OF_None))
     return EC;
@@ -1090,6 +1117,7 @@
   auto Result = md5_contents(FD);
   close(FD);
   return Result;
+#endif
 }
 
 bool exists(const basic_file_status &status) {
@@ -1255,10 +1283,12 @@
 
 Error TempFile::discard() {
   Done = true;
+#if !defined(__wasm__)
   if (FD != -1 && close(FD) == -1) {
     std::error_code EC = errnoAsErrorCode();
     return errorCodeToError(EC);
   }
+#endif
   FD = -1;
 
 #ifdef _WIN32
@@ -1325,6 +1355,7 @@
   if (!RenameEC)
     TmpName = "";
 
+#if !defined(__wasm__)
   if (close(FD) == -1)
     return errorCodeToError(errnoAsErrorCode());
+#endif
   FD = -1;
 
   return errorCodeToError(RenameEC);
@@ -1345,9 +1376,11 @@
 
   TmpName = "";
 
+#if !defined(__wasm__)
   if (close(FD) == -1)
     return errorCodeToError(errnoAsErrorCode());
+#endif
   FD = -1;
 
   return Error::success();
